### ISSUE TITLE
Implement target derivative for smooth L1 loss

### ISSUE THREAD

**Issue by phizaz** [pytorch/pytorch#12327]
## 🚀 Feature
Implement target derivative for `F.smooth_l1_loss`

## Motivation
I'm implementing an actor-critic algorithm. On the TD update step, I need gradients through both the input and the target of the loss, notably both `v(s_t)` and `v(s_t+1)`, without which it won't work properly. I could use `F.mse_loss` for the moment. However, I still want to compare these two losses. 

## Pitch
`F.smooth_l1_loss(input, target)` where both `input` and `target` can have `requires_grad`.

**Comment by fmassa** [comment]
In the meantime, you can implement `smooth_l1_loss` via the following:
```python
def smooth_l1_loss(input, target, beta=1, size_average=True):
    """
    very similar to the smooth_l1_loss from pytorch, but with
    the extra beta parameter
    """
    n = torch.abs(input - target)
    cond = n < beta
    loss = torch.where(cond, 0.5 * n ** 2 / beta, n - 0.5 * beta)
    if size_average:
        return loss.mean()
    return loss.sum()
```
